### DISCUSSION TITLE
An odd case of a fired "binding" action and a `enum` State

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-composable-architecture#1377] [Q&A] [answered]
(I'm using the protocol beta branch with this).

It takes a few steps to explain the setup...

## First

I've created a fairly small app that has an enum at the top level to drive which "section" of the app the user is in.

```
enum State {
  case verifying(Verify.State)
  case login(Login.State)
  case mainApp(MainApp.State)
}
```

## Second
In the Login feature I use two `@BindableState` one for the user phone number and one for the PIN number...

```
public struct State: Equatable {
  @BindableState var phoneNumber: String = ""
  @BindableState var pinNumber: String = ""
}
```

## Third
I also have an action in the `Login` feature `pinButtonTapped`. Which calls an effect that hits an endpoint. This has a possible response action of `pinSuccess`.

## Fourth
The view then brings these together...

```
Section("PINNumber") {
  TextField("123456", text: viewStore.binding(\.$pinNumber))

  Button {
    viewStore.send(.pinButtonTapped)
  } label: {
    Text("Validate PIN")
  }
}
```

## Finally
In the `AppCore` I'm connecting this together and responding to the login...

```
public var body: some ReducerProtocol<State, Action> {
  Reduce { state, action in
    switch action {
    // other cases...
    case .login(. pinSuccess):
      state = .mainApp(.init())
      return .none

    case .login:
      return .none

    // other cases
  }
  .ifCaseLet(/State.login, action: /Action.login) {
    LoginFeature()
  }
  .ifCaseLet(/State.mainApp, action: /Action.mainApp) {
    MainAppFeature()
  }
  .ifCaseLet(/State.verifying, action: /Action.verifying) {
    VerifyingFeature()
  }
}
```

## The problem
When I tap the "Validate PIN" button it sends the action. Then the back end is called. Then the success response is received and the above reducer switches to the `mainApp` state. All of this is fine. But...

If the user still has the PIN number field then it seems like SwiftUI is implicitly updating the binding when the view is dismiss. This triggers two purple warnings. One for the `binding` action not being processed. And another for the `AppCore` receiving an action for state that does not exist.

I tried to manually set the `FocusedField` to nil before doing the auth stuff but then I got another error that `publishing changes from within view updates is not allowed`.

I tried another update about using a `@Binding` and synchronising manually but that didn't work either.

Not sure where to go with this next.

**Comment by oliverfoggin** [comment, marked as answer]
OK... I'm fairly certain this is a bug with SwiftUI/Xcode 14.

From here there is a post near the bottom that links to a video...

https://developer.apple.com/forums/thread/711899

By adding `.buttonStyle(BorderlessButtonStyle())` to the buttons the error about publishing changes goes away.

Which is really odd...

**Reply by tgrapperon** [reply]
If your `Button` is in the cell of a `List` or a `Form`, you're likely hitting #1319. By specifying a `buttonStyle` explicitly, you're opting-out from automatic `List`/`Form` row-tap promotion and the button then behaves like a normal one (this is the standard trick for when you want a button to be tappable in a `List` without triggering the selection).
The issue is then probably the same as #1319 where these taps from the `List/Form` are delivered at the wrong moment.
